CRM/Sqltasks/Action/CallTask.php

```diff
@@ -50,48 +51,140 @@ public function getDefaultOrder() {
-    if (empty($tasks) && empty($categories)) {
-      return;
-    }
+    $task_ids = $this->getConfigValue('tasks') ?? [];
+    $categories = $this->getConfigValue('categories') ?? [];
+    $execute_disabled_tasks = $this->getConfigValue('is_execute_disabled_tasks') == 1;
 
-    // generate query for task selection
-    $query = "SELECT * FROM `civicrm_sqltasks` WHERE  ";
-    $query .= " archive_date IS NULL AND ";
-    if (!$isExecuteDisabledTasks) {
-      $query .= " enabled=1 AND ";
-    }
+    $execute_in_parallel =
+      $this->getConfigValue('execute_in_parallel') == 1
+      && self::backgroundQueueEnabled();
 
-    $or_clauses = array();
-    if (!empty($tasks)) {
-      $or_clauses[] = '`id` IN (' . implode(',', $tasks) . ')';
-    }
-    if (!empty($categories)) {
-      $escaped_categories = array();
-      foreach ($categories as $category) {
-        $escaped_categories[] = "'" . CRM_Core_DAO::escapeString($category) . "'";
-      }
-      $or_clauses[] = '`category` IN (' . implode(',', $escaped_categories) . ')';
-    }
-    $query .= '((' . implode(') OR (', $or_clauses). '))';
-    $query .= ' ORDER BY weight ASC';
-    $tasks2run = CRM_Sqltasks_Task::getTasks($query);
+    if (empty($task_ids) && empty($categories)) return;
+
+    $tasks_2_run = self::findTasks([
+      'categories'             => $categories,
+      'execute_disabled_tasks' => $execute_disabled_tasks,
+      'task_ids'               => $task_ids,
+    ]);
 
-    foreach ($tasks2run as $task) {
-      if ($task->getID() == $this->task->getID()) {
+    $queued_execs = [];
+    $error_count = 0;
+
+    foreach ($tasks_2_run as $task) {
+      $task_id = $task->getID();
+      $task_name = $task->getAttribute('name');
+
+      if ($task_id == $this->task->getID()) continue;
+
+      if ($execute_in_parallel) {
+        $queued_execs[] = $task->executeAsync()['execution_id'];
+        $this->log("Queued task '$task_name' [$task_id] for execution");
         continue;
       }
-      // all good: execute!
-      $taskExecutionResult = $task->execute();
-      $logs = $taskExecutionResult['logs'];
-      // log results from child task
-      foreach ($logs as $log) {
+
+      $exec_result = $task->execute();
+
+      foreach ($exec_result['logs'] as $log) {
         $this->log($log);
       }
-      $this->log("Executed task '" . $task->getAttribute('name') . "' [" . $task->getID() . ']');
+
+      $this->log("Executed task '$task_name' [$task_id]");
     }
+
+    while (!empty($queued_execs)) {
+      $queued_execs_copy = $queued_execs;
+
+      foreach ($queued_execs_copy as $i => $exec_id) {
+        if (self::executionHasEnded($exec_id)) {
+          unset($queued_execs[$i]);
+          $error_count += self::getErrorCount($exec_id);
+        }
+      }
+
+      sleep(1);
+    }
+
+    if ($error_count > 0) {
+      throw new Exception("Execution of task $task_id encountered errors.");
```

Comment:
`$task_id` is misleading here since it's just the last value seen in the `foreach` above, not necessarily the task that encountered an error.

Suggestion: Move the log line within `if (self::executionHasEnded($exec_id)) {` above.

It would be nice if we could also use `$exec_id` to get the task log once it's finished, and add it to the current task log (the way it's done for synchronous execution in line 88).